### PR TITLE
Fix pane info in BigQuery load job id

### LINKED ISSUE #28219: [Bug]: BigQuery IO Batch load using File_load causing the same job id ignoring inserts as the job_id is already completed
### What happened?

```
BigQueryIO.Write<Row> batchWrite =
          rowWrite
              .withMethod(BigQueryIO.Write.Method.FILE_LOADS)
              .withTriggeringFrequency(Duration.standardMinutes(fileLoadSpecOptional.get().writeTriggerFrequency))
              .withAutoSharding()
              .withExtendedErrorInfo()
              .withMaxRetryJobs(fileLoadSpecOptional.get().withJobRetryCount)
              .withWriteTempDataset(TMP_TABLE_DATASET)
              .withCustomGcsTempLocation(ValueProvider.StaticValueProvider.of(fileLoadSpecOptional.get().stagingBucket));
```

Have following code to setup file_load operation for Dataflow streaming jobs using GCS and File_notification based ingestion into BigQuery
- It creates a JobId for new file after few minutes and load data into Live table  with jobId `beam_bq_job_LOAD_defaultnetworkvpcflowlogslive_e78561b1b6e147e7abafe83b9314c7f1_41dd2dbb7ce5b6e717e5aadb4244b4e0_00001_00000-0`
 but after this every new files are added to staging bucket new set of files have exactly `1` partition and c.pane.index seems to be `0` causing the same job_id and ignoring file writes
 See the screenshot for different timestamp we have the same job_id ignoring inserts
image

Should we add random uuid to jobId logic to fix this issue [here](https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java#L250) ?

by changing this line to:
```
 String jobIdPrefix =
          BigQueryResourceNaming.createJobIdWithDestination(
              c.sideInput(loadJobIdPrefixView), tableDestination, partition, c.pane().getIndex()) + "-" +  random5CharUUID();
```

### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by Abacn**
Fixes #28219

* Add randomness to job id

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by Abacn** [comment]
Run PostCommit_Java_Dataflow

**Comment by Abacn** [comment]
Run PostCommit_Java_DataflowV2

**Comment by Abacn** [comment]
There are still multiple code paths relies on the pane info, e.g. write_disposition and create_disposition settings: https://github.com/apache/beam/blob/3ff66d38c41fde475f71254d889ba46440904238/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteRename.java#L194C67-L194C67

let me think more about it. Probably one can save the pane index in WriteResult also

**Review by yeshvantbhavnasi — APPROVED** [review]

**Comment by Abacn** [comment]
there were test failures due to periodic impulse timing. Re-designed test. If passed then ready to review

**Comment by Abacn** [comment]
Run PostCommit_Java_DataflowV2

**Comment by Abacn** [comment]
Although Dataflow PostCommit have multiple quota exceeded failure, bigquery tests all passed: https://ci-beam.apache.org/job/beam_PostCommit_Java_DataflowV2_PR/218/testReport/org.apache.beam.sdk.io.gcp.bigquery/

**Comment by Abacn** [comment]
there are multiple quota exceeded error for post commit, however bigquery tests all passed: https://ci-beam.apache.org/job/beam_PostCommit_Java_DataflowV2_PR/218/testReport/org.apache.beam.sdk.io.gcp.bigquery/

**Comment by Abacn** [comment]
R: @ahmedabu98

will fix spotless in next iteration

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/PeriodicImpulse.java
Comment (ahmedabu98) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/PeriodicImpulse.java:86]: mark with `@Internal`?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryResourceNaming.java
Comment (ahmedabu98) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryResourceNaming.java:63]: I'm worried this may cause duplication of data due to bundle retry. 

Let's say a bundle fails during load/copy execution and the BQ job was successful. Beam would process the bundle again and these lines will create a fresh job ID. Under previous circumstances, this job ID would be recognized by BQ as a recently successful job and will ignore it. But now since the ID is always new, it will execute the job and we will end up with duplicate data.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteIT.java
Comment (ahmedabu98) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteIT.java:106]: Nice, this fixes the test cases that were supposed to be streaming but weren't.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteRename.java
Comment (ahmedabu98) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteRename.java:211]: Unrelated to this PR but we're actually seeing a bug due to relying on the pane index in this code path: #28309

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/PeriodicImpulse.java
Comment (ahmedabu98) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/PeriodicImpulse.java:86]: P.S. also add to `catchUpToNow()`

Comment (Abacn) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/PeriodicImpulse.java:88]: added

**Comment by Abacn** [comment]
Run PostCommit_Java_DataflowV2

**Comment by Abacn** [comment]
dataflow v2 postcommit passed; java precommit known flink test flakiness. PTAL @ahmedabu98

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java
Comment (ahmedabu98) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java:105]: Why are we maintaining pane index coming out of write tables? I don't see that we're using it anywhere downstream

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteRename.java
Comment (Abacn) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteRename.java:222]: the destination name (without hash) may have PII. In the past I put the logs that may contain PII to debug level so they do not get stored by default (unless override to debug).

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java
Comment (Abacn) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java:105]: will revert the WriteTable.Result changes. Originally it was just to keep WritePartition.Result and WriteTable.Result similar.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteRename.java
Comment (Abacn) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteRename.java:222]: (note: final table destination is hased to jobId)

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java
Comment (Abacn) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java:180]: I am going to preserve the change to PendingJobData, as the pane index info would be helpful to resolve possible racing condition of CREATE_TRUNCATE pending jobs

Comment (ahmedabu98) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java:180]: Do you mean for debugging purposes? Will note that `paneIndex()` is still not being called anywhere

**Comment by ahmedabu98** [comment]
Run PostCommit_Java_Dataflow

**Comment by ahmedabu98** [comment]
Run PostCommit_Java_DataflowV2

**Review by ahmedabu98 — APPROVED** [review]
LGTM if tests pass

**Comment by kennknowles** [comment]
Are you sure this is right? Isn't the purpose of the job id to prevent duplicates?

**Comment by Abacn** [comment]
@kennknowles thanks for confirm. The latest approach is deterministic now (thanks the comment by @ahmedabu98), the WritePartition.Result now preserves the pane index before the ReShuffle happens is used afterwards. The behavior should be totally the same as before on Dataflow Legacy runner, while fixing Dataflow V2.

**Comment by Abacn** [comment]
Java PreCommit known flakiness due to #28383 - might need to disable these tests again

**Comment by Abacn** [comment]
Run PostCommit_Java_DataflowV2

**Comment by reuvenlax** [comment]
What exactly is this trying to fix? As @kennknowles we do rely on the job id to prevent duplicates (basically we can sometimes reissue a load job and rely on BigQuery to reject the duplicate loads because it has the same job id). We also want to avoid update-incompatible changes here.

**Comment by ahmedabu98** [comment]
After a Reshuffle (we have one [before single loads](https://github.com/ahmedabu98/beam/blob/67e88d9fec9b4ac39a6d42a4bc1e64e1236bc309/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java#L806) and [temp loads](https://github.com/ahmedabu98/beam/blob/67e88d9fec9b4ac39a6d42a4bc1e64e1236bc309/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java#L763)), RunnerV2 doesn't propagate pane indices like legacy runner does. Instead of an increasing pane index, RunnerV2 keeps it at 0. 

After our reshuffles, we directly rely on pane index to [construct a job ID](https://github.com/ahmedabu98/beam/blob/67e88d9fec9b4ac39a6d42a4bc1e64e1236bc309/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java#L252-L254). In the RunnerV2 case the index stays at 0 so in some cases (e.g. single loads) it produces the same job ID for each trigger. The resulting behavior is that only the first load is successful and the rest are seen as duplicates by BQ and ignored. 

Yi's solution here is to save the pane index from WritePartitions (a stage that is before Reshuffle) and use that information instead when creating the Job ID in WriteTables.

**Comment by reuvenlax** [comment]
On Mon, Sep 11, 2023 at 9:02 AM Ahmed Abualsaud ***@***.***>
wrote:

> After a Reshuffle (we have one before single loads
> <https://github.com/ahmedabu98/beam/blob/67e88d9fec9b4ac39a6d42a4bc1e64e1236bc309/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java#L806>
> and temp loads
> <https://github.com/ahmedabu98/beam/blob/67e88d9fec9b4ac39a6d42a4bc1e64e1236bc309/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java#L763>),
> RunnerV2 doesn't propagate pane indices like legacy runner does. Instead of
> an increasing pane index, RunnerV2 keeps it at 0.
>

This sounds like a bug in Runner V2. Pane indices are supposed to propagate
through shuffles, and we have various things that depend on this.


> After our reshuffles, we directly rely on pane index to construct a job ID
> <https://github.com/ahmedabu98/beam/blob/67e88d9fec9b4ac39a6d42a4bc1e64e1236bc309/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java#L252-L254>.
> In the RunnerV2 case the index stays at 0 so in some cases (e.g. single
> loads) it produces the same job ID for each trigger. The resulting behavior
> is that only the first load is successful and the rest are seen as
> duplicates by BQ and ignored.
>

Should we simply file a bug against RunnerV2?

**Comment by kennknowles** [comment]
Actually each GBK has its own sequence of trigger firings, and when it comes to reshuffle there are two possibilities:

1. The reshuffle does a GBK then explodes it (per its expansion) in which case the sequence of pane indices come from the Reshuffle/always trigger on that GBK.
2. The reshuffle is a semantics "no op" at the PCollection level, in which case the pane indices would be preserved from an upstream aggregation, but reshuffle results in bundles from random upstream keys so they are deceptive.

**Comment by kennknowles** [comment]
I don't see a way for pane indices crossing a reshuffle to make any sense.

**Comment by Abacn** [comment]
Hi @reuvenlax @kennknowles is there an agreement on the fix?

In any case, we identified there are other issues in FILE_LOAD streaming (#28309) and a fix (#28312) involves changing the coder (and element type) for another GBK used in the pipeline, so we have to break the update compatibility anyways. I can make sure this fix get in as the same beam version as #28312 and note them in CHANGES.md

**Comment by reuvenlax** [comment]
I'm not sure what you mean by random upstream keys. Reshuffle
currently maps KV -> KV, so no keys are changed. The only difference is
that the keys are now spread out on different workers, which is a semantic
no op.

You may be thinking of Rehsuffle.viaRandomKey, which is a bit different.

**Comment by kennknowles** [comment]
Right. I was, indeed, thinking of reshuffle via random key. But what I'm saying does apply to Reshuffle by assigned keys. I have been a person who always says that Reshuffle is a semantic noop. If we should preserve that we need the v2 and the SDK implementation to record the pane info (aka reify `WindowedValue` instead of `PaneIndex`). This issue has shaken up my beliefs.

If you have `Combine -> (stage) -> Reshuffle`. Each input to reshuffle has a pane index that corresponds to its key+window in the combine. The main thing is that if you have the original key+window+index you can reassemble the aggregate and/or write it to a sink in a consistent way. But if the stage changes the key you lose that property. Do we even have a good story for what happens to pane info when a DoFn reassigns to new keys?

**Comment by kennknowles** [comment]
Notably we only have tests like `ReshuffleTest.testReshufflePreservesTimestamps` https://github.com/apache/beam/blob/cf0cf3b746c55921ebe4a95d2e06052cb96be74d/sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ReshuffleTest.java#L115

Considering that reifying and testing against a full windowed value would have be somewhat of a default thing to do, I would guess the original intent was very aware of this issue.

**Comment by Abacn** [comment]
Run PostCommit_Java_DataflowV2

**Comment by Abacn** [comment]
Tests now exercised and passed on Dataflow Runner V2: https://ci-beam.apache.org/job/beam_PostCommit_Java_DataflowV2_PR/226/testReport/org.apache.beam.sdk.io.gcp.bigquery/FileLoadsStreamingIT/

**Comment by ahmedabu98** [comment]
@reuvenlax as @Abacn mentioned, we broke update-compatibility recently in #28312 so that we can unblock dynamic destinations with copy jobs (#28309). Do we have the green light to merge these changes? 
If Dataflow folks decide Runner V2 should preserve pane indices, that support may still take some time to implement. In the meantime, this ensures file loads streaming works on the runner (IIUC it previously never did)

**Comment by reuvenlax** [comment]
FYI - https://issues.apache.org/jira/browse/BEAM-7195 was caused by the
inverse issue. The old code assumed that Reshuffle would produce pane
indices, but of course it just forwards them.

Let's wait until we figure out the correct behavior. Very few users will be
exercising this from runner v2 today, so I would rather be cautious here -
especially if the right answer is to fix runner v2.

**Comment by kennknowles** [comment]
Yea let's sync on this. I think I've come around but I want to get it on dev@ in at least a 1 pager for what the spec is. If it turns out to not be dangerous to reinstate, and we can regain the semantic property "reshuffle is a semantic noop" then that is pretty desirable, so that runners can freely add and remove reshuffles.

Noting that here is where Dataflow v1 implemented reshuffle and re-instated the pre-reshuffle pane: https://github.com/apache/beam/blob/34024902746af90e7bf41e28729ec031dbab58d2/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingGroupAlsoByWindowReshuffleFn.java#L61

Noting that Beam's reference implementation, the closest thing we have to a spec, does not preserve the pane but will produce a sequence of increasing panes.

**Comment by Abacn** [comment]
@kennknowles thanks for pin to the related code path for Dataflow legacy worker. I now understand that Dataflow legacy work does some special PTransform override for ReShuffle. For portable pipelines, is that true that there is no longer similar PTransform override, and a possible fix would change java core's ReShuffle to also recover pane (currently it recovers Timestamp) ?
https://github.com/apache/beam/blob/346541b1b824e0343a736ce4e320667b8d73ec06/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java#L105

**Comment by kennknowles** [comment]
I expect there is a similar override inside the Dataflow service and Unified Worker. The implementation in the SDK would be slow and just for reference, if it was executed. A reshuffle really should not have to cross the Fn API.

**Comment by kennknowles** [comment]
But yes if we make the SDK recover the pane info also, then that would make the SDK (direct runner, prism, etc) match Dataflow v1. This would probably not fix v2, but it would fix "the spec".

**Comment by kennknowles** [comment]
OK so when I thought about it I agree with everyone else that the reference implementation is wrong and that the right behavior is to be the same as the identity function on PCollections. That means keeping the pane info exactly. I wrote the dev list about it. I think it is probably such a boring thread that it will not get a lot of comment.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java
Comment (kennknowles) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java:259]: My concern is that if we have this:

    GBK -> processing -> reshuffle -> this DoFn

Can you tell me which case we are in?

1. The `processing` left the key the same so the reshuffle is just for checkpoint
2. The `processing` changed the key so that the current element key + pane index is no longer unique

Comment (Abacn) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java:259]: Well it's a little bit more subtle. It is the same key, but became ShardedKey after `processing`

After GBK, key is DestinationT
After processing, key is ShardedKey<DestinationT> (shard is added in WritePartitions)

**Pull request closed without merging**
